File: scripts/extract-official-brand.mjs
/**
 * Extract official Magnifico logo from PDF and sample brand colors.
 * Usage: node scripts/extract-official-brand.mjs [path-to-pdf]
 */
import sharp from "sharp";
import { execFileSync } from "node:child_process";
import { copyFile, mkdir, readFile, writeFile } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const root = path.join(__dirname, "..");
const pdfPath =
  process.argv[2] ??
  "C:\\Users\\Wissam Sbaity\\Downloads\\magnifico.pdf";
const publicDir = path.join(root, "public");
const brandingDir = path.join(publicDir, "branding");
const logoOut = path.join(publicDir, "logo.png");

function rgbToHex(r, g, b) {
  return `#${[r, g, b].map((v) => v.toString(16).padStart(2, "0")).join("")}`;
}

function colorDistance(a, b) {
  return Math.hypot(a[0] - b[0], a[1] - b[1], a[2] - b[2]);
}

function clusterColors(pixels, k = 8) {
  const samples = pixels.filter((p) => p[3] > 200);
  if (!samples.length) return [];

  const centroids = [];
  for (let i = 0; i < k; i++) {
    const p = samples[(Math.random() * samples.length) | 0];
    centroids.push([p[0], p[1], p[2]]);
  }

  for (let iter = 0; iter < 12; iter++) {
    const buckets = Array.from({ length: k }, () => []);
    for (const p of samples) {
      let best = 0;
      let bestDist = Infinity;
      for (let i = 0; i < k; i++) {
        const d = colorDistance(p, centroids[i]);
        if (d < bestDist) {
          bestDist = d;
          best = i;
        }
      }
      buckets[best].push(p);
    }
    for (let i = 0; i < k; i++) {
      if (!buckets[i].length) continue;
      centroids[i] = buckets[i]
        .reduce((acc, p) => [acc[0] + p[0], acc[1] + p[1], acc[2] + p[2]], [0, 0, 0])
        .map((v) => Math.round(v / buckets[i].length));
    }
  }

  const counts = Array(k).fill(0);
  for (const p of samples) {
    let best = 0;
    let bestDist = Infinity;
    for (let i = 0; i < k; i++) {
      const d = colorDistance(p, centroids[i]);
      if (d < bestDist) {
        bestDist = d;
        best = i;
      }
    }
    counts[best]++;
  }

  return centroids
    .map((rgb, i) => ({ rgb, count: counts[i] }))
    .filter((c) => c.count > 0)
    .sort((a, b) => b.count - a.count);
}

function classifyBrandColors(clusters) {
  const scored = clusters.map(({ rgb, count }) => {
    const [r, g, b] = rgb;
    const max = Math.max(r, g, b);
    const min = Math.min(r, g, b);
    const sat = max === 0 ? 0 : (max - min) / max;
    const lum = (0.2126 * r + 0.7152 * g + 0.0722 * b) / 255;
    return { rgb, count, sat, lum };
  });

  const yellow = scored
    .filter((c) => c.rgb[0] > 180 && c.rgb[1] > 120 && c.rgb[2] < 120 && c.sat > 0.25)
    .sort((a, b) => b.count - a.count)[0];
  const pink = scored
    .filter((c) => c.rgb[0] > 150 && c.rgb[1] < 130 && c.rgb[2] > 60 && c.rgb[0] > c.rgb[1])
    .sort((a, b) => b.count - a.count)[0];
  const green = scored
    .filter((c) => c.rgb[1] > c.rgb[0] && c.rgb[1] > c.rgb[2] && c.rgb[1] > 90)
    .sort((a, b) => b.count - a.count)[0];
  const black = scored
    .filter((c) => c.lum < 0.2)
    .sort((a, b) => b.count - a.count)[0];
  const cream = scored
    .filter((c) => c.lum > 0.85 && c.sat < 0.2)
    .sort((a, b) => b.count - a.count)[0];

  return { yellow, pink, green, black, cream, all: scored };
}

async function renderPdfToPng(inputPdf, outputPng) {
  try {
    execFileSync(
      "magick",
      ["-density", "300", `${inputPdf}[0]`, "-background", "none", outputPng],
      { stdio: "pipe" }
    );
    return true;
  } catch {
    return false;
  }
}

async function trimTransparentLogo(inputPath, outputPath) {
  await sharp(inputPath)
    .trim({ threshold: 12 })
    .png({ compressionLevel: 9, adaptiveFiltering: true })
    .toFile(outputPath);
  const meta = await sharp(outputPath).metadata();
  return meta;
}

async function sampleColors(imagePath) {
  const { data, info } = await sharp(imagePath)
    .ensureAlpha()
    .raw()
    .toBuffer({ resolveWithObject: true });
  const pixels = [];
  for (let i = 0; i < data.length; i += info.channels) {
    pixels.push([data[i], data[i + 1], data[i + 2], data[i + 3]]);
  }
  return classifyBrandColors(clusterColors(pixels, 10));
}

async function main() {
  await mkdir(brandingDir, { recursive: true });

  const workingPng = path.join(brandingDir, "official-logo-source.png");
  let rendered = await renderPdfToPng(pdfPath, workingPng);

  if (!rendered) {
    // Fallback: if PDF is already a raster image wrapped in PDF, copy existing high-res source
    console.warn("ImageMagick unavailable — copy PDF manually or install magick.");
    process.exitCode = 1;
    return;
  }

  const meta = await trimTransparentLogo(workingPng, logoOut);
  const colors = await sampleColors(logoOut);

  const tokens = {
    logo: {
      path: "/logo.png",
      width: meta.width,
      height: meta.height,
      sourcePdf: pdfPath,
    },
    colors: {
      yellow: colors.yellow?.rgb ?? null,
      pink: colors.pink?.rgb ?? null,
      green: colors.green?.rgb ?? null,
      black: colors.black?.rgb ?? null,
      cream: colors.cream?.rgb ?? null,
    },
    hex: Object.fromEntries(
      Object.entries(colors)
        .filter(([k]) => k !== "all")
        .map(([k, v]) => [k, v?.rgb ? rgbToHex(...v.rgb) : null])
    ),
  };

  await writeFile(
    path.join(root, "src", "lib", "branding", "brand-tokens.json"),
    JSON.stringify(tokens, null, 2)
  );

  console.log(JSON.stringify(tokens, null, 2));
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: scripts/extract-official-brand.py
"""Extract logo + brand colors from official Magnifico PDF."""
from __future__ import annotations

import json
import os
import sys
from collections import Counter

try:
    import fitz  # pymupdf
except ImportError:
    import subprocess

    subprocess.check_call([sys.executable, "-m", "pip", "install", "pymupdf", "-q"])
    import fitz

from PIL import Image
import io

PDF = r"C:\Users\Wissam Sbaity\Downloads\magnifico.pdf"
ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
PUBLIC = os.path.join(ROOT, "public")
BRANDING = os.path.join(PUBLIC, "branding")
LOGO_OUT = os.path.join(PUBLIC, "logo.png")
TOKENS_OUT = os.path.join(ROOT, "src", "lib", "branding", "brand-tokens.json")

os.makedirs(BRANDING, exist_ok=True)


def rgb_to_hex(rgb):
    return "#{:02x}{:02x}{:02x}".format(*rgb)


def sample_colors(img: Image.Image):
    img = img.convert("RGBA")
    pixels = [p for p in img.getdata() if p[3] > 200]
    # quantize to reduce noise
    buckets = Counter()
    for r, g, b, a in pixels:
        key = (r // 8 * 8, g // 8 * 8, b // 8 * 8)
        buckets[key] += 1

    ranked = buckets.most_common(40)

    def pick(filter_fn):
        for rgb, count in ranked:
            if filter_fn(rgb):
                return {"rgb": list(rgb), "count": count, "hex": rgb_to_hex(rgb)}
        return None

    yellow = pick(lambda c: c[0] > 170 and c[1] > 110 and c[2] < 130 and c[0] > c[2])
    pink = pick(lambda c: c[0] > 150 and c[1] < 130 and c[2] > 60 and c[0] > c[1])
    green = pick(lambda c: c[1] > c[0] + 15 and c[1] > c[2] + 15 and c[1] > 90)
    black = pick(lambda c: max(c) < 55)
    cream = pick(lambda c: min(c) > 230 and max(c) - min(c) < 20)

    return {
        "yellow": yellow,
        "pink": pink,
        "green": green,
        "black": black,
        "cream": cream,
        "top": [{"rgb": list(k), "hex": rgb_to_hex(k), "count": v} for k, v in ranked[:12]],
    }


def main():
    doc = fitz.open(PDF)
    page = doc[0]
    img_path = None
    best_img = None

    for img in page.get_images(full=True):
        xref = img[0]
        base = doc.extract_image(xref)
        if not best_img or base["width"] * base["height"] > best_img["width"] * best_img["height"]:
            best_img = base

    if best_img:
        ext = best_img["ext"]
        img_path = os.path.join(BRANDING, f"official-logo-source.{ext}")
        with open(img_path, "wb") as f:
            f.write(best_img["image"])
        pil = Image.open(io.BytesIO(best_img["image"]))
    else:
        mat = fitz.Matrix(4, 4)
        pix = page.get_pixmap(matrix=mat, alpha=True)
        img_path = os.path.join(BRANDING, "official-logo-source.png")
        pix.save(img_path)
        pil = Image.open(img_path)

    doc.close()

    # Trim transparent margins
    bbox = pil.getbbox()
    if bbox:
        pil = pil.crop(bbox)

    pil.save(LOGO_OUT, format="PNG", optimize=True)
    w, h = pil.size
    colors = sample_colors(pil)

    tokens = {
        "logo": {"path": "/logo.png", "width": w, "height": h, "sourcePdf": PDF},
        "colors": {k: (colors[k]["rgb"] if colors.get(k) else None) for k in ["yellow", "pink", "green", "black", "cream"]},
        "hex": {k: (colors[k]["hex"] if colors.get(k) else None) for k in ["yellow", "pink", "green", "black", "cream"]},
        "samples": colors["top"],
    }

    with open(TOKENS_OUT, "w", encoding="utf-8") as f:
        json.dump(tokens, f, indent=2)

    print(json.dumps(tokens, indent=2))


if __name__ == "__main__":
    main()
